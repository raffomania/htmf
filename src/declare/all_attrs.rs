use std::borrow::Cow;

use crate::attr::Attr;
use crate::element::Element;

// Take care to name the parameter `value`
// to disable rust analyzer inlay hints
macro_rules! define_attr_function {
    ($name:ident) => {
        pub fn $name<'a, C>(value: C) -> Attr<'a>
        where
            C: Into<Cow<'a, str>>,
        {
            Attr(stringify!($name), value.into())
        }
    };
    ($name:ident, $key:literal) => {
        pub fn $name<'a, C>(value: C) -> Attr<'a>
        where
            C: Into<Cow<'a, str>>,
        {
            Attr($key, value.into())
        }
    };
    ($name:ident, $key:literal, $value:literal) => {
        pub fn $name<'a>() -> Attr<'a> {
            Attr($key, $value.into())
        }
    };
}

macro_rules! define_attr_method {
    ($name:ident) => {
        pub fn $name<C>(mut self, value: C) -> Element<'a>
        where
            C: Into<Cow<'a, str>>,
        {
            if let Some(attrs) = self.element.attrs_mut() {
                attrs.push($name(value));
            }
            self
        }
    };
    ($name:ident, $value:literal) => {
        pub fn $name<C>(mut self) -> Element<'a> {
            if let Some(attrs) = self.element.attrs_mut() {
                attrs.push($name());
            }
            self
        }
    };
}

impl<'a> Element<'a> {
    define_attr_method!(accept);
    define_attr_method!(accept_charset);
    define_attr_method!(action);
    define_attr_method!(alt);
    define_attr_method!(aria_checked);
    define_attr_method!(aria_current);
    define_attr_method!(aria_disabled, "true");
    define_attr_method!(aria_hidden, "true");
    define_attr_method!(aria_invalid, "true");
    define_attr_method!(aria_label);
    define_attr_method!(aria_labelledby);
    define_attr_method!(aria_placeholder);
    define_attr_method!(aria_readonly);
    define_attr_method!(aria_required);
    define_attr_method!(async_, "true");
    define_attr_method!(autocapitalize);
    define_attr_method!(autocomplete);
    define_attr_method!(autofocus, "true");
    define_attr_method!(autoplay, "true");
    define_attr_method!(capture);
    define_attr_method!(charset);
    define_attr_method!(checked, "true");
    define_attr_method!(cite_attr);
    define_attr_method!(class);
    define_attr_method!(content);
    define_attr_method!(contenteditable, "true");
    define_attr_method!(crossorigin);
    define_attr_method!(defer, "true");
    define_attr_method!(disabled, "true");
    define_attr_method!(draggable, "true");
    define_attr_method!(enctype);
    define_attr_method!(for_);
    define_attr_method!(formaction);
    define_attr_method!(height);
    define_attr_method!(href);
    define_attr_method!(http_equiv);
    define_attr_method!(id);
    define_attr_method!(integrity);
    define_attr_method!(lang);
    define_attr_method!(loop_, "true");
    define_attr_method!(maxlength);
    define_attr_method!(method);
    define_attr_method!(minlength);
    define_attr_method!(name);
    define_attr_method!(placeholder);
    define_attr_method!(preload, "true");
    define_attr_method!(property);
    define_attr_method!(readonly, "true");
    define_attr_method!(rel);
    define_attr_method!(required);
    define_attr_method!(role);
    define_attr_method!(selected, "true");
    define_attr_method!(src);
    define_attr_method!(style);
    define_attr_method!(tabindex);
    define_attr_method!(target);
    define_attr_method!(title);
    define_attr_method!(type_);
    define_attr_method!(value);
    define_attr_method!(width);
}

define_attr_function!(accept);
define_attr_function!(accept_charset);
define_attr_function!(action);
define_attr_function!(alt);
define_attr_function!(aria_checked);
define_attr_function!(aria_current);
define_attr_function!(aria_disabled, "aria_disabled", "true");
define_attr_function!(aria_hidden, "aria_hidden", "true");
define_attr_function!(aria_invalid, "aria_invalid", "true");
define_attr_function!(aria_label);
define_attr_function!(aria_labelledby);
define_attr_function!(aria_placeholder);
define_attr_function!(aria_readonly);
define_attr_function!(aria_required);
define_attr_function!(async_, "async", "true");
define_attr_function!(autocapitalize);
define_attr_function!(autocomplete);
define_attr_function!(autofocus, "autofocus", "true");
define_attr_function!(autoplay, "autoplay", "true");
define_attr_function!(capture);
define_attr_function!(charset);
define_attr_function!(checked, "checked", "true");
define_attr_function!(cite_attr);
define_attr_function!(class);
define_attr_function!(content);
define_attr_function!(contenteditable, "contenteditable", "true");
define_attr_function!(crossorigin);
define_attr_function!(defer, "defer", "true");
define_attr_function!(disabled, "disabled", "true");
define_attr_function!(draggable, "draggable", "true");
define_attr_function!(enctype);
define_attr_function!(for_, "for");
define_attr_function!(formaction);
define_attr_function!(height);
define_attr_function!(href);
define_attr_function!(http_equiv);
define_attr_function!(id);
define_attr_function!(integrity);
define_attr_function!(lang);
define_attr_function!(loop_, "loop", "true");
define_attr_function!(maxlength);
define_attr_function!(method);
define_attr_function!(minlength);
define_attr_function!(name);
define_attr_function!(placeholder);
define_attr_function!(preload, "preload", "true");
define_attr_function!(property);
define_attr_function!(readonly, "readonly", "true");
define_attr_function!(rel);
define_attr_function!(required);
define_attr_function!(role);
define_attr_function!(selected, "selected", "true");
define_attr_function!(src);
define_attr_function!(style);
define_attr_function!(tabindex);
define_attr_function!(target);
define_attr_function!(title);
define_attr_function!(type_, "type");
define_attr_function!(value);
define_attr_function!(width);
