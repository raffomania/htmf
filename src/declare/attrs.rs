use std::borrow::Cow;

use crate::Attr;

macro_rules! define_attr_function {
    ($name:ident) => {
        pub fn $name<'a, C>(content: C) -> Attr<'a>
        where
            C: Into<Cow<'a, str>>,
        {
            (stringify!($name), content.into())
        }
    };
    ($name:ident, $value:literal) => {
        pub fn $name<'a>() -> Attr<'a> {
            (stringify!($name), $value.into())
        }
    };
}

define_attr_function!(accept);
define_attr_function!(accept_charset);
define_attr_function!(action);
define_attr_function!(alt);
define_attr_function!(aria_checked);
define_attr_function!(aria_current);
define_attr_function!(aria_disabled, "true");
define_attr_function!(aria_hidden, "true");
define_attr_function!(aria_invalid, "true");
define_attr_function!(aria_label);
define_attr_function!(aria_labelledby);
define_attr_function!(aria_placeholder);
define_attr_function!(aria_readonly);
define_attr_function!(aria_required);
define_attr_function!(async_, "true");
define_attr_function!(autocapitalize);
define_attr_function!(autocomplete);
define_attr_function!(autofocus, "true");
define_attr_function!(autoplay, "true");
define_attr_function!(capture);
define_attr_function!(charset);
define_attr_function!(checked, "true");
define_attr_function!(cite_attr);
define_attr_function!(class);
define_attr_function!(content);
define_attr_function!(contenteditable, "true");
define_attr_function!(crossorigin);
define_attr_function!(defer, "true");
define_attr_function!(disabled, "true");
define_attr_function!(draggable, "true");
define_attr_function!(enctype);
define_attr_function!(for_);
define_attr_function!(formaction);
define_attr_function!(height);
define_attr_function!(href);
define_attr_function!(http_equiv);
define_attr_function!(id);
define_attr_function!(integrity);
define_attr_function!(lang);
define_attr_function!(loop_, "true");
define_attr_function!(maxlength);
define_attr_function!(method);
define_attr_function!(minlength);
define_attr_function!(name);
define_attr_function!(placeholder);
define_attr_function!(preload, "true");
define_attr_function!(property);
define_attr_function!(readonly, "true");
define_attr_function!(rel);
define_attr_function!(required);
define_attr_function!(role);
define_attr_function!(selected, "true");
define_attr_function!(src);
define_attr_function!(style);
define_attr_function!(tabindex);
define_attr_function!(target);
define_attr_function!(title);
define_attr_function!(type_);
define_attr_function!(value);
define_attr_function!(width);
