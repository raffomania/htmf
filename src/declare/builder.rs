use crate::Element;

pub struct Builder<'borrowed, 'element> {
    pub element: &'borrowed mut Element<'element>,
}

impl<'borrowed, 'element> Builder<'borrowed, 'element> {
    pub fn add_children<'temp: 'borrowed, C>(self, children: C) -> Self
    where
        C: Into<Vec<Element<'element>>>,
    {
        let mut children = children.into();
        self.element.children_mut().append(&mut children);
        self
    }
}
