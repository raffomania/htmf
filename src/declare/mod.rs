mod all_attrs;
mod all_tags;

use std::borrow::Cow;

pub use crate::element::Element;
pub use all_attrs::*;
pub use all_tags::*;

use crate::element::{Path, PureElement};

pub fn text<'e, C>(value: C) -> Element<'e>
where
    C: Into<Cow<'e, str>>,
{
    Element {
        element: PureElement::Text { text: value.into() },
        parent: Path::Top,
    }
}

impl<'e> Element<'e> {
    pub fn text<C>(self, value: C) -> Element<'e>
    where
        C: Into<Cow<'e, str>>,
    {
        self.into_new_child_element(PureElement::Text { text: value.into() })
    }
}

pub fn fragment<'e>() -> Element<'e> {
    Element {
        element: PureElement::Fragment {
            children: Vec::new(),
        },
        parent: Path::Top,
    }
}

/// Prepend `<!doctype html>` to the given children.
pub fn document<'e>() -> Element<'e> {
    Element {
        element: PureElement::Document {
            children: Vec::new(),
        },
        parent: Path::Top,
    }
}
