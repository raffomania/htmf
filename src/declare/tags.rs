use crate::Element;

macro_rules! define_tag_function {
    ($tag:ident $(, leaf)*) => {
        pub fn $tag<'a>() -> Element<'a> {
            Element::Tag {
                tag: stringify!($tag),
                attrs: Vec::new(),
                children: Vec::new(),
            }
        }
    };

    ($tag:ident, $tag_str:literal) => {
        pub fn $tag<'a>() -> Element<'a> {
            Element::Tag {
                tag: $tag_str,
                attrs: Vec::new(),
                children: Vec::new(),
            }
        }
    };
}

/// Prepend `<!doctype html>` to the given children.
pub fn document<'a>() -> Element<'a> {
    Element::Document {
        children: Vec::new(),
    }
}

macro_rules! define_tag_method {
    ($tag:ident $(, leaf)*) => {
        pub fn $tag<'b>(&'b mut self) -> &'b mut Element<'a> {
            self.children_mut().push($tag());
            self.children_mut().last_mut().unwrap()
        }
    };
}

impl<'a> Element<'a> {
    pub(crate) fn children_mut<'b>(&'b mut self) -> &'b mut Vec<Element<'a>> {
        match self {
            Element::Tag {
                tag: _,
                attrs: _,
                ref mut children,
            } => children,
            Element::Fragment { ref mut children } => children,
            Element::Text(_) => todo!(),
            Element::Document { ref mut children } => children,
        }
    }

    define_tag_method!(a);
    define_tag_method!(abbr);
    define_tag_method!(address);
    define_tag_method!(area, leaf);
    define_tag_method!(article);
    define_tag_method!(aside);
    define_tag_method!(audio);
    define_tag_method!(b);
    define_tag_method!(base, leaf);
    define_tag_method!(bdi);
    define_tag_method!(bdo);
    define_tag_method!(body);
    define_tag_method!(blockquote);
    define_tag_method!(br, leaf);
    define_tag_method!(button);
    define_tag_method!(canvas);
    define_tag_method!(caption);
    define_tag_method!(cite);
    define_tag_method!(code);
    define_tag_method!(col);
    define_tag_method!(colgroup);
    define_tag_method!(data);
    define_tag_method!(datalist);
    define_tag_method!(dd);
    define_tag_method!(del);
    define_tag_method!(details);
    define_tag_method!(dfn);
    define_tag_method!(dialog);
    define_tag_method!(div);
    define_tag_method!(dl);
    define_tag_method!(dt);
    define_tag_method!(em);
    define_tag_method!(embed);
    define_tag_method!(fieldset);
    define_tag_method!(figcaption);
    define_tag_method!(figure);
    define_tag_method!(footer);
    define_tag_method!(form);
    define_tag_method!(h1);
    define_tag_method!(h2);
    define_tag_method!(h3);
    define_tag_method!(h4);
    define_tag_method!(h5);
    define_tag_method!(h6);
    define_tag_method!(header);
    define_tag_method!(head);
    define_tag_method!(hr, leaf);
    define_tag_method!(html);
    define_tag_method!(i);
    define_tag_method!(iframe);
    define_tag_method!(img, leaf);
    define_tag_method!(input, leaf);
    define_tag_method!(ins);
    define_tag_method!(kbd);
    define_tag_method!(label);
    define_tag_method!(legend);
    define_tag_method!(li);
    define_tag_method!(link, leaf);
    define_tag_method!(main_);
    define_tag_method!(map);
    define_tag_method!(mark);
    define_tag_method!(math);
    define_tag_method!(menu);
    define_tag_method!(menuitem);
    define_tag_method!(meta, leaf);
    define_tag_method!(meter);
    define_tag_method!(nav);
    define_tag_method!(noscript);
    define_tag_method!(object);
    define_tag_method!(ol);
    define_tag_method!(optgroup);
    define_tag_method!(option);
    define_tag_method!(output);
    define_tag_method!(p);
    define_tag_method!(param);
    define_tag_method!(picture);
    define_tag_method!(pre);
    define_tag_method!(progress);
    define_tag_method!(q);
    define_tag_method!(rp);
    define_tag_method!(rt);
    define_tag_method!(ruby);
    define_tag_method!(s);
    define_tag_method!(samp);
    define_tag_method!(script);
    define_tag_method!(section);
    define_tag_method!(select);
    define_tag_method!(small);
    define_tag_method!(source, leaf);
    define_tag_method!(span);
    define_tag_method!(strong);
    define_tag_method!(sub);
    define_tag_method!(summary);
    define_tag_method!(sup);
    define_tag_method!(svg);
    define_tag_method!(table);
    define_tag_method!(tbody);
    define_tag_method!(td);
    define_tag_method!(textarea);
    define_tag_method!(tfoot);
    define_tag_method!(th);
    define_tag_method!(thead);
    define_tag_method!(time);
    define_tag_method!(tr);
    define_tag_method!(track, leaf);
    define_tag_method!(u);
    define_tag_method!(ul);
    define_tag_method!(var);
    define_tag_method!(video);
    define_tag_method!(wbr);
}

define_tag_function!(a);
define_tag_function!(abbr);
define_tag_function!(address);
define_tag_function!(area, leaf);
define_tag_function!(article);
define_tag_function!(aside);
define_tag_function!(audio);
define_tag_function!(b);
define_tag_function!(base, leaf);
define_tag_function!(bdi);
define_tag_function!(bdo);
define_tag_function!(body);
define_tag_function!(blockquote);
define_tag_function!(br, leaf);
define_tag_function!(button);
define_tag_function!(canvas);
define_tag_function!(caption);
define_tag_function!(cite);
define_tag_function!(code);
define_tag_function!(col);
define_tag_function!(colgroup);
define_tag_function!(data);
define_tag_function!(datalist);
define_tag_function!(dd);
define_tag_function!(del);
define_tag_function!(details);
define_tag_function!(dfn);
define_tag_function!(dialog);
define_tag_function!(div);
define_tag_function!(dl);
define_tag_function!(dt);
define_tag_function!(em);
define_tag_function!(embed);
define_tag_function!(fieldset);
define_tag_function!(figcaption);
define_tag_function!(figure);
define_tag_function!(footer);
define_tag_function!(form);
define_tag_function!(h1);
define_tag_function!(h2);
define_tag_function!(h3);
define_tag_function!(h4);
define_tag_function!(h5);
define_tag_function!(h6);
define_tag_function!(header);
define_tag_function!(head);
define_tag_function!(hr, leaf);
define_tag_function!(html);
define_tag_function!(i);
define_tag_function!(iframe);
define_tag_function!(img, leaf);
define_tag_function!(input, leaf);
define_tag_function!(ins);
define_tag_function!(kbd);
define_tag_function!(label);
define_tag_function!(legend);
define_tag_function!(li);
define_tag_function!(link, leaf);
define_tag_function!(main_, "main");
define_tag_function!(map);
define_tag_function!(mark);
define_tag_function!(math);
define_tag_function!(menu);
define_tag_function!(menuitem);
define_tag_function!(meta, leaf);
define_tag_function!(meter);
define_tag_function!(nav);
define_tag_function!(noscript);
define_tag_function!(object);
define_tag_function!(ol);
define_tag_function!(optgroup);
define_tag_function!(option);
define_tag_function!(output);
define_tag_function!(p);
define_tag_function!(param);
define_tag_function!(picture);
define_tag_function!(pre);
define_tag_function!(progress);
define_tag_function!(q);
define_tag_function!(rp);
define_tag_function!(rt);
define_tag_function!(ruby);
define_tag_function!(s);
define_tag_function!(samp);
define_tag_function!(script);
define_tag_function!(section);
define_tag_function!(select);
define_tag_function!(small);
define_tag_function!(source, leaf);
define_tag_function!(span);
define_tag_function!(strong);
define_tag_function!(sub);
define_tag_function!(summary);
define_tag_function!(sup);
define_tag_function!(svg);
define_tag_function!(table);
define_tag_function!(tbody);
define_tag_function!(td);
define_tag_function!(textarea);
define_tag_function!(tfoot);
define_tag_function!(th);
define_tag_function!(thead);
define_tag_function!(time);
define_tag_function!(tr);
define_tag_function!(track, leaf);
define_tag_function!(u);
define_tag_function!(ul);
define_tag_function!(var);
define_tag_function!(video);
define_tag_function!(wbr);
