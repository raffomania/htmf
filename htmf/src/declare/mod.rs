pub(crate) mod all_attrs;
mod all_tags;

use crate::{
    attr::{Attr, Attrs},
    element::Element,
};

#[cfg(feature = "unstable-builder")]
pub use crate::builder::Builder;
pub use all_attrs::*;
pub use all_tags::*;

#[cfg(feature = "unstable-builder")]
mod builder {
    use crate::builder::Builder;
    use crate::element::Element;

    impl Builder {
        pub fn text<C>(self, value: C) -> Builder
        where
            C: Into<String>,
        {
            self.into_new_child_element(Element::Text { text: value.into() })
        }
    }
}

pub fn text<C>(value: C) -> Element
where
    C: Into<String>,
{
    Element::Text { text: value.into() }
}

/// Prepend `<!doctype html>` to the given children.
pub fn document() -> Element {
    Element::Document {
        children: Vec::new(),
    }
}

pub fn fragment() -> Element {
    Element::Fragment {
        children: Vec::new(),
    }
}

pub fn nothing() -> Element {
    Element::Nothing
}

pub fn attr<C>(name: &'static str, value: C) -> Attrs
where
    C: Into<String>,
{
    Attrs(vec![Attr(name, value.into())])
}
