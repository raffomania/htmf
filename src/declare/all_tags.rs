use crate::attr::IntoAttrs;

#[cfg(feature = "unstable-builder")]
use crate::builder::Builder;

use crate::element::Element;

macro_rules! define_tag_function {
    ($tag:ident $(, leaf)*) => {
        pub fn $tag<Attrs: IntoAttrs>(value: Attrs) -> Element {
            Element::Tag {
                tag: stringify!($tag),
                attrs: value.into_attrs(),
                children: Vec::new(),
            }
        }
    };

    ($tag:ident, $tag_str:literal) => {
        pub fn $tag<Attrs: IntoAttrs>(value: Attrs) -> Element {
            Element::Tag {
                tag: $tag_str,
                attrs: value.into_attrs(),
                children: Vec::new(),
            }
        }
    };
}

#[cfg(feature = "unstable-builder")]
macro_rules! define_tag_builder_method {
    ($tag:ident $(, leaf)*) => {
        pub fn $tag<Attrs: IntoAttrs>(self, value: Attrs) -> Builder {
            self.into_new_child_tag(stringify!($tag), value.into_attrs())
        }
    };
    ($tag:ident, $tag_str:literal) => {
        pub fn $tag<Attrs: IntoAttrs>(self, value: Attrs) -> Builder {
            self.into_new_child_tag($tag_str, value.into_attrs())
        }
    };
}

#[cfg(feature = "unstable-builder")]
macro_rules! define_tag_element_method {
    ($tag:ident $(, leaf)*) => {
        pub fn $tag<Attrs: IntoAttrs>(self, value: Attrs) -> Builder {
            Builder::from(self).into_new_child_tag(stringify!($tag), value.into_attrs())
        }
    };
    ($tag:ident, $tag_str:literal) => {
        pub fn $tag<Attrs: IntoAttrs>(self, value: Attrs) -> Builder {
            Builder::from(self).into_new_child_tag($tag_str, value.into_attrs())
        }
    };
}

#[cfg(feature = "unstable-builder")]
impl Builder {
    define_tag_builder_method!(a);
    define_tag_builder_method!(abbr);
    define_tag_builder_method!(address);
    define_tag_builder_method!(area, leaf);
    define_tag_builder_method!(article);
    define_tag_builder_method!(aside);
    define_tag_builder_method!(audio);
    define_tag_builder_method!(b);
    define_tag_builder_method!(base, leaf);
    define_tag_builder_method!(bdi);
    define_tag_builder_method!(bdo);
    define_tag_builder_method!(body);
    define_tag_builder_method!(blockquote);
    define_tag_builder_method!(br, leaf);
    define_tag_builder_method!(button);
    define_tag_builder_method!(canvas);
    define_tag_builder_method!(caption);
    define_tag_builder_method!(cite);
    define_tag_builder_method!(code);
    define_tag_builder_method!(col);
    define_tag_builder_method!(colgroup);
    define_tag_builder_method!(data);
    define_tag_builder_method!(datalist);
    define_tag_builder_method!(dd);
    define_tag_builder_method!(del);
    define_tag_builder_method!(details);
    define_tag_builder_method!(dfn);
    define_tag_builder_method!(dialog);
    define_tag_builder_method!(div);
    define_tag_builder_method!(dl);
    define_tag_builder_method!(dt);
    define_tag_builder_method!(em);
    define_tag_builder_method!(embed);
    define_tag_builder_method!(fieldset);
    define_tag_builder_method!(figcaption);
    define_tag_builder_method!(figure);
    define_tag_builder_method!(footer);
    define_tag_builder_method!(form);
    define_tag_builder_method!(h1);
    define_tag_builder_method!(h2);
    define_tag_builder_method!(h3);
    define_tag_builder_method!(h4);
    define_tag_builder_method!(h5);
    define_tag_builder_method!(h6);
    define_tag_builder_method!(header);
    define_tag_builder_method!(head);
    define_tag_builder_method!(hr, leaf);
    define_tag_builder_method!(html);
    define_tag_builder_method!(i);
    define_tag_builder_method!(iframe);
    define_tag_builder_method!(img, leaf);
    define_tag_builder_method!(input, leaf);
    define_tag_builder_method!(ins);
    define_tag_builder_method!(kbd);
    define_tag_builder_method!(label);
    define_tag_builder_method!(legend);
    define_tag_builder_method!(li);
    define_tag_builder_method!(link, leaf);
    define_tag_builder_method!(main_, "main");
    define_tag_builder_method!(map);
    define_tag_builder_method!(mark);
    define_tag_builder_method!(math);
    define_tag_builder_method!(menu);
    define_tag_builder_method!(menuitem);
    define_tag_builder_method!(meta, leaf);
    define_tag_builder_method!(meter);
    define_tag_builder_method!(nav);
    define_tag_builder_method!(noscript);
    define_tag_builder_method!(object);
    define_tag_builder_method!(ol);
    define_tag_builder_method!(optgroup);
    define_tag_builder_method!(option);
    define_tag_builder_method!(output);
    define_tag_builder_method!(p);
    define_tag_builder_method!(param);
    define_tag_builder_method!(picture);
    define_tag_builder_method!(pre);
    define_tag_builder_method!(progress);
    define_tag_builder_method!(q);
    define_tag_builder_method!(rp);
    define_tag_builder_method!(rt);
    define_tag_builder_method!(ruby);
    define_tag_builder_method!(s);
    define_tag_builder_method!(samp);
    define_tag_builder_method!(script);
    define_tag_builder_method!(section);
    define_tag_builder_method!(select);
    define_tag_builder_method!(small);
    define_tag_builder_method!(source, leaf);
    define_tag_builder_method!(span);
    define_tag_builder_method!(strong);
    define_tag_builder_method!(sub);
    define_tag_builder_method!(summary);
    define_tag_builder_method!(sup);
    define_tag_builder_method!(svg);
    define_tag_builder_method!(table);
    define_tag_builder_method!(tbody);
    define_tag_builder_method!(td);
    define_tag_builder_method!(textarea);
    define_tag_builder_method!(tfoot);
    define_tag_builder_method!(th);
    define_tag_builder_method!(thead);
    define_tag_builder_method!(time);
    define_tag_builder_method!(tr);
    define_tag_builder_method!(track, leaf);
    define_tag_builder_method!(u);
    define_tag_builder_method!(ul);
    define_tag_builder_method!(var);
    define_tag_builder_method!(video);
    define_tag_builder_method!(wbr);
}

#[cfg(feature = "unstable-builder")]
impl Element {
    define_tag_element_method!(a);
    define_tag_element_method!(abbr);
    define_tag_element_method!(address);
    define_tag_element_method!(area, leaf);
    define_tag_element_method!(article);
    define_tag_element_method!(aside);
    define_tag_element_method!(audio);
    define_tag_element_method!(b);
    define_tag_element_method!(base, leaf);
    define_tag_element_method!(bdi);
    define_tag_element_method!(bdo);
    define_tag_element_method!(body);
    define_tag_element_method!(blockquote);
    define_tag_element_method!(br, leaf);
    define_tag_element_method!(button);
    define_tag_element_method!(canvas);
    define_tag_element_method!(caption);
    define_tag_element_method!(cite);
    define_tag_element_method!(code);
    define_tag_element_method!(col);
    define_tag_element_method!(colgroup);
    define_tag_element_method!(data);
    define_tag_element_method!(datalist);
    define_tag_element_method!(dd);
    define_tag_element_method!(del);
    define_tag_element_method!(details);
    define_tag_element_method!(dfn);
    define_tag_element_method!(dialog);
    define_tag_element_method!(div);
    define_tag_element_method!(dl);
    define_tag_element_method!(dt);
    define_tag_element_method!(em);
    define_tag_element_method!(embed);
    define_tag_element_method!(fieldset);
    define_tag_element_method!(figcaption);
    define_tag_element_method!(figure);
    define_tag_element_method!(footer);
    define_tag_element_method!(form);
    define_tag_element_method!(h1);
    define_tag_element_method!(h2);
    define_tag_element_method!(h3);
    define_tag_element_method!(h4);
    define_tag_element_method!(h5);
    define_tag_element_method!(h6);
    define_tag_element_method!(header);
    define_tag_element_method!(head);
    define_tag_element_method!(hr, leaf);
    define_tag_element_method!(html);
    define_tag_element_method!(i);
    define_tag_element_method!(iframe);
    define_tag_element_method!(img, leaf);
    define_tag_element_method!(input, leaf);
    define_tag_element_method!(ins);
    define_tag_element_method!(kbd);
    define_tag_element_method!(label);
    define_tag_element_method!(legend);
    define_tag_element_method!(li);
    define_tag_element_method!(link, leaf);
    define_tag_element_method!(main_, "main");
    define_tag_element_method!(map);
    define_tag_element_method!(mark);
    define_tag_element_method!(math);
    define_tag_element_method!(menu);
    define_tag_element_method!(menuitem);
    define_tag_element_method!(meta, leaf);
    define_tag_element_method!(meter);
    define_tag_element_method!(nav);
    define_tag_element_method!(noscript);
    define_tag_element_method!(object);
    define_tag_element_method!(ol);
    define_tag_element_method!(optgroup);
    define_tag_element_method!(option);
    define_tag_element_method!(output);
    define_tag_element_method!(p);
    define_tag_element_method!(param);
    define_tag_element_method!(picture);
    define_tag_element_method!(pre);
    define_tag_element_method!(progress);
    define_tag_element_method!(q);
    define_tag_element_method!(rp);
    define_tag_element_method!(rt);
    define_tag_element_method!(ruby);
    define_tag_element_method!(s);
    define_tag_element_method!(samp);
    define_tag_element_method!(script);
    define_tag_element_method!(section);
    define_tag_element_method!(select);
    define_tag_element_method!(small);
    define_tag_element_method!(source, leaf);
    define_tag_element_method!(span);
    define_tag_element_method!(strong);
    define_tag_element_method!(sub);
    define_tag_element_method!(summary);
    define_tag_element_method!(sup);
    define_tag_element_method!(svg);
    define_tag_element_method!(table);
    define_tag_element_method!(tbody);
    define_tag_element_method!(td);
    define_tag_element_method!(textarea);
    define_tag_element_method!(tfoot);
    define_tag_element_method!(th);
    define_tag_element_method!(thead);
    define_tag_element_method!(time);
    define_tag_element_method!(tr);
    define_tag_element_method!(track, leaf);
    define_tag_element_method!(u);
    define_tag_element_method!(ul);
    define_tag_element_method!(var);
    define_tag_element_method!(video);
    define_tag_element_method!(wbr);
}

define_tag_function!(a);
define_tag_function!(abbr);
define_tag_function!(address);
define_tag_function!(area, leaf);
define_tag_function!(article);
define_tag_function!(aside);
define_tag_function!(audio);
define_tag_function!(b);
define_tag_function!(base, leaf);
define_tag_function!(bdi);
define_tag_function!(bdo);
define_tag_function!(body);
define_tag_function!(blockquote);
define_tag_function!(br, leaf);
define_tag_function!(button);
define_tag_function!(canvas);
define_tag_function!(caption);
define_tag_function!(cite);
define_tag_function!(code);
define_tag_function!(col);
define_tag_function!(colgroup);
define_tag_function!(data);
define_tag_function!(datalist);
define_tag_function!(dd);
define_tag_function!(del);
define_tag_function!(details);
define_tag_function!(dfn);
define_tag_function!(dialog);
define_tag_function!(div);
define_tag_function!(dl);
define_tag_function!(dt);
define_tag_function!(em);
define_tag_function!(embed);
define_tag_function!(fieldset);
define_tag_function!(figcaption);
define_tag_function!(figure);
define_tag_function!(footer);
define_tag_function!(form);
define_tag_function!(h1);
define_tag_function!(h2);
define_tag_function!(h3);
define_tag_function!(h4);
define_tag_function!(h5);
define_tag_function!(h6);
define_tag_function!(header);
define_tag_function!(head);
define_tag_function!(hr, leaf);
define_tag_function!(html);
define_tag_function!(i);
define_tag_function!(iframe);
define_tag_function!(img, leaf);
define_tag_function!(input, leaf);
define_tag_function!(ins);
define_tag_function!(kbd);
define_tag_function!(label);
define_tag_function!(legend);
define_tag_function!(li);
define_tag_function!(link, leaf);
define_tag_function!(main_, "main");
define_tag_function!(map);
define_tag_function!(mark);
define_tag_function!(math);
define_tag_function!(menu);
define_tag_function!(menuitem);
define_tag_function!(meta, leaf);
define_tag_function!(meter);
define_tag_function!(nav);
define_tag_function!(noscript);
define_tag_function!(object);
define_tag_function!(ol);
define_tag_function!(optgroup);
define_tag_function!(option);
define_tag_function!(output);
define_tag_function!(p);
define_tag_function!(param);
define_tag_function!(picture);
define_tag_function!(pre);
define_tag_function!(progress);
define_tag_function!(q);
define_tag_function!(rp);
define_tag_function!(rt);
define_tag_function!(ruby);
define_tag_function!(s);
define_tag_function!(samp);
define_tag_function!(script);
define_tag_function!(section);
define_tag_function!(select);
define_tag_function!(small);
define_tag_function!(source, leaf);
define_tag_function!(span);
define_tag_function!(strong);
define_tag_function!(sub);
define_tag_function!(summary);
define_tag_function!(sup);
define_tag_function!(svg);
define_tag_function!(table);
define_tag_function!(tbody);
define_tag_function!(td);
define_tag_function!(textarea);
define_tag_function!(tfoot);
define_tag_function!(th);
define_tag_function!(thead);
define_tag_function!(time);
define_tag_function!(tr);
define_tag_function!(track, leaf);
define_tag_function!(u);
define_tag_function!(ul);
define_tag_function!(var);
define_tag_function!(video);
define_tag_function!(wbr);
