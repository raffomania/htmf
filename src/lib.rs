pub mod attr;
pub mod declare;
pub mod element;

#[cfg(test)]
mod tests {
    use super::*;

    use declare::*;

    use pretty_assertions::assert_eq;

    #[test]
    fn base() {
        let doc_immutable = document()
            .with([html().class("w-full h-full").with([
                head().with([
                    link().rel("stylesheet").href("/assets/preflight.css"),
                    link().rel("stylesheet").href("/assets/railwind.css"),
                    script().src("/assets/htmx.1.9.9.js"),
                    meta().name("color-scheme").content("dark"),
                    meta()
                        .name("viewport")
                        .content("width=device-width,initial-scale=1"),
                ]),
                body().class("w-full h-full text-gray-200 bg-neutral-800"),
            ])])
            .build();
        let html_immutable = doc_immutable.clone().to_html();
        insta::assert_snapshot!(html_immutable);

        let doc_mut = document()
            .html()
            .class("w-full h-full")
            .with([head().with([
                link().rel("stylesheet").href("/assets/preflight.css"),
                link().rel("stylesheet").href("/assets/railwind.css"),
                script().src("/assets/htmx.1.9.9.js"),
                meta().name("color-scheme").content("dark"),
                meta()
                    .name("viewport")
                    .content("width=device-width,initial-scale=1"),
            ])])
            .body()
            .class("w-full h-full text-gray-200 bg-neutral-800")
            .build();
        let html_mutable = doc_mut.to_html();
        assert_eq!(html_immutable, html_mutable);
        assert_eq!(doc_immutable, doc_mut);
    }
}
