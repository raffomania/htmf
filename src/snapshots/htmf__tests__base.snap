---
source: src/lib.rs
expression: html_immutable
---
<!doctype html>

<html class="w-full h-full">
    <head>
        <link rel="stylesheet" href="/assets/preflight.css"></link>
        <link rel="stylesheet" href="/assets/railwind.css"></link>
        <script src="/assets/htmx.1.9.9.js"></script>
        <meta name="color-scheme" content="dark"></meta>
        <meta name="viewport" content="width=device-width,initial-scale=1"></meta>
    </head>
    <body class="w-full h-full text-gray-200 bg-neutral-800"></body>
</html>
